### PR TITLE
Add: Token Learner for Vision Transformer

### PR DESCRIPTION
**Pull request by ariG23498**
In this example we (with @sayakpaul) have implemented [TokenLearner: What Can 8 Learned Tokens Do for Images and Videos?](https://arxiv.org/abs/2106.11297) with a smaller ViT for CIFAR-10 image classification. 

In the paper the authors suggest an adaptive token learning algorithm that makes ViT computationally much more efficient (in terms of FLOPs) and also increases downstream accuracy (here classification accuracy). Experimenting with CIFAR-10 we reduce the number of pathces from 64 to 4 (number of adaptively learned tokens) and also report a boost in the accuracy.

We think this should be a great addition to the keras examples as it would be accessible to the community for further research.

Colab Link: https://colab.research.google.com/drive/1zjAnHQ3LLfS47dxTqaWm77YAa7WljTXo?usp=sharing

### CODE REVIEW

File: examples/vision/token_learner.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        x = layers.Dropout(dropout_rate)(x)
+    return x
+
+
+"""
+## TokenLearner module
+
+The following figure presents a pictorial overview of the module
+([source](https://ai.googleblog.com/2021/12/improving-vision-transformer-efficiency.html)).
+([source](https://ai.googleblog.com/2021/12/improving-vision-transformer-efficiency.html)).
```
Comment (sayakpaul) [inline examples/vision/token_learner.py:212]: @ariG23498 needs fix.

Comment (ariG23498) [reply examples/vision/token_learner.py]: Fixed.

File: examples/vision/token_learner.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        layers.Rescaling(1 / 255.0),
+        layers.Resizing(INPUT_SHAPE[0] + 20, INPUT_SHAPE[0] + 20),
+        layers.RandomCrop(IMAGE_SIZE, IMAGE_SIZE),
+        layers.RandomFlip("horizontal"),
+    ],
+    name="data_augmentation",
+)
+
+"""
+Note that these layers have their inference behaviour pre-defined. Refer
```
Comment (fchollet) [inline examples/vision/token_learner.py:158]: The meaning of the sentence "these layers have their inference behaviour pre-defined" is unclear. Do this refer to the fact that augmentation layers (above, that's `RandomCrop` and `RandomFlip`) are pass-through when `training=False`?

File: examples/vision/token_learner.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+"""
+## Results
+
+We experimented with and without the TokenLearner inside the mini ViT we implemented
+(with the same hyperparameters presented in this example). Here are our results:
+
+| **TokenLearner** | **# tokens in<br> TokenLearner** | **Top-1 Acc<br>(Averaged across 5 runs)** | **GFLOPs** | **TensorBoard** |
+|:---:|:---:|:---:|:---:|:---:|
+| N | - | 56.112% | 0.0184 | [Link](https://tensorboard.dev/experiment/vkCwM49dQZ2RiK0ZT4mj7w/) |
```
Comment (fchollet) [inline examples/vision/token_learner.py:446]: Very neat use of Tensorboard.dev!

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/vision/token_learner.py (hunk 2 above)
Comment (sayakpaul) [reply examples/vision/token_learner.py:158]: That's correct. However, RandomCrop's inference behavior is a little different. From the [official docs](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_crop/):

> At inference time, the images will be first rescaled to preserve the shorter side, and center cropped.

**Comment by sayakpaul** [comment]
@fchollet thank you for your comments. Just addressed them.

File: examples/vision/token_learner.py [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+
+"""
+## Imports
+"""
+
+import tensorflow as tf
+
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_addons as tfa
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/token_learner.py:64]: That would be a great addition.

**Review by fchollet — APPROVED** [review]
Wonderful. I pushed some copyedits -- please pull them first, then you can add the generated files. Thank you!

**Comment by sayakpaul** [comment]
@fchollet done. Thank you! Ready for you to take over. 

Also, we have added GFLOP counts for each all the models to show the compute effectivity of the TokenLearner module.

**Review by fchollet — APPROVED** [review]
Merging -- thank you for the great contribution! 👍
